stm32-modules/heater-shaker/include/heater-shaker/gcode_parser.hpp

```diff
@@ -0,0 +1,228 @@
+/*
+** gcode_parser - a templated gcode parser for use anywhere
+*/
+#pragma once
+
+#include <stdio.h>
+
+#include <algorithm>
+#include <cctype>
+#include <charconv>
+#include <iterator>
+#include <optional>
+#include <string>
+#include <utility>
+#include <variant>
+
+namespace gcode {
```

Comment:
Just so I have this straight in my head:
- This is just a template for generic gcode parsing. We will probably need to add more specific implementation for the heater shaker and that would live in a cpp file, right?

Reply:
No not really; what will happen is the host comms task core will use the individual gcode parsers implemented in gcode.hpp and instantiate the template there:

```
void HostCommsTask::handle_message(const Message& message) {
    ....
    auto parser = gcode::GroupParser<gcode::M3, gcode::M115, gcode::...>();

    auto result = std::make_pair(parser::ParseResult(std::monostate), message.cbegin());
    while (true) {
      result = parser.parse_available(result.second, message.cend());
      if (std::holds_alternative<std::monostate>(result.first()) {
          break;
      } else {
         build_and_dispatch(result.first);
      }
    } 
```

or something similar